### DISCUSSION TITLE
Setup modules dependency

### DISCUSSION THREAD

**Discussion by chesti84** [statiqdev/Discussions#178] [Q&A]
Hi,

I have a pipeline defined with a lot of `ProcessModules`. I need to create another process module that will run after `ProcessShortcodes`.
[I have read the documentation](https://www.statiq.dev/guide/pipelines/modules/module-patterns#concatenation), and it seems to be possible to do it, but I cannot find a way to set things up.

Can you help me with this?

**Comment by nils-a** [comment]
I /think/

* *Modules* can only be in parent-child relationships, like when you derive something from [ParentModule](https://github.com/statiqdev/Statiq.Framework/blob/1a8bd58e6f1a4e63f0cae2870db78126bc511928/src/core/Statiq.Common/Modules/ParentModule.cs).
* Dependencies can be built only between *Pipelines*, as described in [Specifying Dependencies](https://www.statiq.dev/guide/pipelines/defining-pipelines#specifying-dependencies)

I could be entirely wrong, though. 😄

**Comment by daveaglick** [comment]
In the answer above, @nils-a wasn't wrong on either count - both the advice on module object-oriented hierarchy and dependencies are correct. However, I think you might be asking a slightly different question. If I'm understanding the question, you want to "insert" a module into an existing pipeline's process phase at a specific position after a `ProcessShortcodes` module that's already in the pipeline. Is that correct?

If so, the easiest way to do it is with the bootstrapper fluent API and the `ModifyPipeline` method with the `InsertAfterFirst()` extension method on the `ProcessModules` collection:

```csharp
await Bootstrapper.Factory
// ...
  .ModifyPipeline(
      "MyPipeline",
      pipeline => pipeline.ProcessModules.InsertAfterFirst<ProcessShortcodes>(new MyModule()))
// ...
```

There are a number of other extensions for module lists like `InsertBeforeFirst<TModule>()`, etc. - but the important thing to keep in mind is that the collection of modules for a given pipeline phase is essentially a list (they actually implement `IList<IModule>`) and can be manipulated as such prior to execution.

It's worth noting that the notion of "dependency" is a little bit of a misnomer when it comes to modules. _Pipelines_ can have dependencies on other pipelines and the engine works out the full dependency chain and throws an error if a circular dependency is discovered. Modules are always sequential within a given pipeline and phase, so module B has a "dependency" on module A if (and only if) module B appears after module A in the sequence of modules for that pipeline and phase.

If this question is about how to make a module in one pipeline "wait" until a module in a different pipeline is complete (I.e. if the `ProcessShortcodes` module you want your module to run after isn't actually in the same pipeline) that's not strictly possible. The way I'd accomplish that is to split the first pipeline that contains your module into two pipelines and make the second start with your module and have a dependency on both the pipeline you split off with the first half of your other modules _and_ the pipeline that contains `ProcessShortcodes`.

Does that answer the question?
